File: src/generator/analysis/models/regions.py
"""Region-specific entity models and clusters.

Following .clinerules standards with modern Python type hints.
"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field

from generator.analysis.models.clusters import BaseEntitiesCluster


# ---- Region entity models ----

class RegionHexTile(BaseModel):
    """Region hex tile entity model."""
    entity_uuid: str = Field(..., description="UUID from filename")
    hex_key: str | None = Field(None, description="Canonical hex key like 'W2S51'")
    map: dict[str, Any] | None = Field(None, description="{'x': float, 'y': float, 'hex_id': str | None}")
    region_uuid: str | None = Field(None, description="UUID of region if distinct")
    settlement_uuids: list[str] = Field(default_factory=list)
    dungeon_uuids: list[str] = Field(default_factory=list)
    faction_uuids: list[str] = Field(default_factory=list)

    @classmethod
    def extract_uuid_connections(cls) -> dict[str, str]:
        """Extract UUID connection types."""
        return {
            "entity_uuid": "region_hex_tile",
            "region_uuid": "region",
            "settlement_uuids": "settlement",
            "dungeon_uuids": "dungeon",
            "faction_uuids": "faction",
        }


# ---- Region cluster specialization ----

class RawRegionEntities(BaseEntitiesCluster):
    """Region-specific cluster for AI model generation."""
    
    def inventory_schema(self) -> dict[str, Any]:
        """Schema for region inventory extraction."""
        return {
            "type": "object",
            "properties": {
                "entities": {
                    "type": "array",
                    "items": {
                        "type": "object",
                        "properties": {
                            "name": {"type": "string"},
                            "description": {"type": "string"},
                            "fields": {
                                "type": "array",
                                "items": {
                                    "type": "object",
                                    "properties": {
                                        "name": {"type": "string"},
                                        "type": {"type": "string"},
                                        "required": {"type": "boolean"},
                                        "description": {"type": "string"},
                                        "is_uuid": {"type": "boolean"},
                                        "is_connection": {"type": "boolean"},
                                    },
                                    "required": ["name", "type", "required"],
                                    "additionalProperties": False,
                                },
                            },
                        },
                        "required": ["name", "fields"],
                        "additionalProperties": False,
                    },
                },
                "connections": {"type": "object", "additionalProperties": {"type": "string"}},
                "notes": {"type": "array", "items": {"type": "string"}},
            },
            "required": ["entities"],
            "additionalProperties": False,
        }

    def analysis_prompt(self) -> str:
        """Prompt for region analysis."""
        return (
            "Analyze the supplied HTML/JSON snippets related to *regions*.\n"
            "Return a JSON object with an 'entities' array describing data models.\n"
            "Focus on names, descriptions, field names/types, and which fields are UUIDs or connections.\n"
            "Look for hex coordinates, map positions, settlements, dungeons, and faction references.\n"
            "If uncertain, omit rather than invent."
        )

    def model_template(self) -> str:
        """Template for region model generation."""
        return '''# Generated Pydantic models for regions
from __future__ import annotations

from typing import Any
from pydantic import BaseModel, Field

{% for e in inventory.entities %}
class {{ e.name }}(BaseModel):
    """{{ e.description or e.name }}"""
{% for f in e.fields %}
    {{ f.name }}: {{ f.type }}{% if not f.required %} | None{% endif %} = {% if f.required %}Field(...{% else %}Field(None{% endif %}, description="{{ f.description or '' }}")
{% endfor %}

    @classmethod
    def extract_uuid_connections(cls) -> dict[str, str]:
        """Extract UUID connection types."""
        return {
{% for f in e.fields if f.is_uuid or f.is_connection %}
            '{{ f.name }}': '{{ inventory.connections.get(f.name, 'unknown') }}',
{% endfor %}
        }

{% endfor %}
'''

File: crates/dl_processors/src/regions.rs
//! Region-specific entity models and processing.
//! 
//! This module contains specialized region processing logic including:
//! - Region entity model matching Python regions.py
//! - RawRegionEntities cluster with specialized AI generation
//! - Region-specific inventory schemas, prompts, and templates

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::Path;
use anyhow::Result;

use dl_analysis::clusters::{BaseEntitiesCluster, EntityCluster};
use dl_types::analysis::raw::{RawEntity, EntityCategory};
use dl_analysis::results::GenerationResults;

/// Specialized cluster for region entities with region-specific AI generation
#[derive(Debug, Clone)]
pub struct RawRegionEntities {
    base: BaseEntitiesCluster,
}

impl RawRegionEntities {
    pub fn new(cluster_name: String) -> Self {
        Self {
            base: BaseEntitiesCluster::new(EntityCategory::Regions, cluster_name),
        }
    }

    pub fn combined() -> Self {
        Self::new("combined".to_string())
    }
}

impl EntityCluster for RawRegionEntities {
    fn category(&self) -> EntityCategory {
        EntityCategory::Regions
    }

    fn cluster_name(&self) -> &str {
        &self.base.cluster_name
    }

    fn add_entity(&mut self, entity: RawEntity) -> bool {
        self.base.add_entity(entity)
    }

    fn can_generate_models(&self) -> bool {
        self.base.can_generate_models()
    }

    fn write_entities_to_disk(&mut self, analysis_dir: &Path) -> Result<()> {
        self.base.write_entities_to_disk(analysis_dir)
    }

    fn inventory_schema(&self) -> serde_json::Value {
        serde_json::json!({
            "type": "object",
            "properties": {
                "entities": {
                    "type": "array",
                    "items": {
                        "type": "object",
                        "properties": {
                            "name": {"type": "string"},
                            "description": {"type": "string"},
                            "fields": {
                                "type": "array",
                                "items": {
                                    "type": "object",
                                    "properties": {
                                        "name": {"type": "string"},
                                        "type": {"type": "string"},
                                        "required": {"type": "boolean"},
                                        "description": {"type": "string"},
                                        "is_uuid": {"type": "boolean"},
                                        "is_connection": {"type": "boolean"}
                                    },
                                    "required": ["name", "type", "required"],
                                    "additionalProperties": false
                                }
                            }
                        },
                        "required": ["name", "fields"],
                        "additionalProperties": false
                    }
                },
                "connections": {"type": "object", "additionalProperties": {"type": "string"}},
                "notes": {"type": "array", "items": {"type": "string"}}
            },
            "required": ["entities"],
            "additionalProperties": false
        })
    }

    fn analysis_prompt(&self) -> String {
        "Analyze the supplied HTML/JSON snippets related to *regions*.\n\
         Return a JSON object with an 'entities' array describing data models.\n\
         Focus on names, descriptions, field names/types, and which fields are UUIDs or connections.\n\
         Look for hex coordinates, map positions, settlements, dungeons, and faction references.\n\
         If uncertain, omit rather than invent.".to_string()
    }

    fn model_template(&self) -> String {
        "regions_analysis.j2".to_string()
    }

    fn generate_models(
        &self,
        models_dir: &Path,
        logger: &mut dyn std::io::Write,
    ) -> Result<GenerationResults> {
        let inventory = self.base.analyze_entities(logger)?;
        self.base.generate_code_from_inventory(&inventory, models_dir, logger)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use dl_types::analysis::raw::RawEntity;

    #[test]
    fn test_region_cluster() {
        let cluster = RawRegionEntities::new("Aurora Bushes".to_string());
        assert_eq!(cluster.category(), EntityCategory::Regions);
        assert_eq!(cluster.cluster_name(), "Aurora Bushes");
    }
}
